Data/HashMap/Base.hs

```diff
@@ -1822,6 +1799,56 @@ updateOrConcatWithKey f ary1 ary2 = A.run $ do
     return mary
 {-# INLINABLE updateOrConcatWithKey #-}
 
+-- Helper functions for very rare cases.
+
+-- Remove (any) one element from an array of hashmaps
+unConsA :: A.Array (HashMap k v) -> UnCons k v
+unConsA ary = case A.length ary of
+  0 -> WasEmpty
+  1 -> case A.index ary 0 of
+    Empty -> WasEmpty -- That would be weird, but fine.
+    Leaf _ l -> NowEmpty l
+    BitmapIndexed bm ary' -> case unConsA ary' of
+      WasEmpty -> WasEmpty -- That would be weird, but fine.
+      NowEmpty l -> NowEmpty l
+      UnConsed l a' -> UnConsed l (A.singleton (BitmapIndexed bm a'))
+    Full ary' -> case unConsA ary' of
+        WasEmpty -> WasEmpty -- That would be weird, but fine.
+        NowEmpty l -> NowEmpty l
+        UnConsed l a' -> UnConsed l (A.singleton (Full a')) -- TODO 'Full' seems wrong
+    Collision h l1 l2 hm -> UnConsed l1 $ A.singleton $ case unConsHM hm of
+      Nothing -> Leaf h l2
+      Just (l3, hm') -> Collision h l2 l3 hm'
+  _ -> goA 0
+  where
+    goA ix = case unConsHM $ A.index ary ix of
+      Nothing -> goA $ ix + 1 -- TODO this is probably problematic when the entire array only houses Empty hashmaps
+      Just (l, hm') -> case hm' of
+        Empty -> UnConsed l $ A.delete ary ix
+        _ -> UnConsed l $ A.update ary ix hm'
+
+data UnCons k v
+  = WasEmpty
+  | NowEmpty (Leaf k v)
+  | UnConsed (Leaf k v) (A.Array (HashMap k v))
+
+-- Remove (any) one element from a hashmap
+unConsHM :: HashMap k v -> Maybe (Leaf k v, HashMap k v)
+unConsHM hm = case hm of
+  Empty -> Nothing
+  Leaf _ l -> Just (l, Empty)
+  BitmapIndexed bm ary' -> case unConsA ary' of
+    WasEmpty -> Nothing
+    NowEmpty l -> Just (l, Empty)
+    UnConsed l a' -> Just (l, BitmapIndexed bm a')
+  Full ary' -> case unConsA ary' of
+    WasEmpty -> Nothing -- That would be weird, but fine.
+    NowEmpty l -> Just (l, Empty)
+    UnConsed l a' -> Just (l, Full a') -- TODO This 'Full' seems wrong
+  Collision h l1 l2 hm' -> Just $ (,) l1 $ case unConsHM hm' of
+    Nothing -> Leaf h l2
+    Just (l3, hm'') -> Collision h l2 l3 hm''
```

Comment:
This smells very wrong. You're pulling up through a collision level without rehashing. I think the right thing here is probably to special-case `hm' = empty` and in all other cases leave `l1` and `l2` alone and "uncons" from the next level down.

Reply:
Er... Sorry.... Ignore this comment. It's bogus.

Reply:
For the record: the comment misses that you don't need to rehash because you already know that the values below will collide on the same hash for this level, because that's the point of the `Collision` constructor.